File: string-paradigm-dependency_2/src/main/java/com/steelzack/string/paradigm/performance2/ParadigmDependency2Impl.java
package com.steelzack.string.paradigm.performance2;

import static com.google.common.base.Preconditions.checkNotNull;

import java.util.List;

import com.steelzack.string.paradigm.api.Dependency;

public class ParadigmDependency2Impl extends Dependency {
	private StringWrapper dependencies;

	public ParadigmDependency2Impl(StringWrapper data) {
		checkNotNull(data);
		this.dependencies = data;
	}

	public int getDependencyCount() {
		return dependencies.toString().split(".").length;
	}

	public String getSubDependencyElement(int index) {
		final String[] splittedValues = dependencies.toString().split(".");
		if (index < splittedValues.length - 1) {
			return splittedValues[index];
		} else {
			return null;
		}
	}

	public boolean startsWith(List<String> subDependencyElement) {
		int dependencyCount = getDependencyCount();
		if (dependencyCount < subDependencyElement.size()) {
			return false;
		}
		for (int i = 0; i < subDependencyElement.size(); i++) {
			if (!getSubDependencyElement(i).equals(subDependencyElement.get(i).toString())) {
				return false;
			}
		}
		return true;
	}
}

File: jeorg-kotlin-apps/string-array-paradigms/string-paradigm-dependency_2/src/main/kotlin/org/jesperancinha/string/paradigm/performance2/ParadigmDependency2Impl.kt
package org.jesperancinha.string.paradigm.performance2

import org.jesperancinha.string.paradigm.api.Dependency

class ParadigmDependency2Impl(
    private val data: StringWrapper
) : Dependency() {

    override fun dependencyCount(): Int {
        return data.toString().split(";".toRegex()).dropLastWhile { it.isEmpty() }.size
    }

    override fun getSubDependencyElement(index: Int): String {
        val splittedValues = data.toString().split(";".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        return if (index < splittedValues.size) {
            splittedValues[index]
        } else {
            ""
        }
    }

    override fun startsWith(subDependencyElement: List<String>): Boolean {
        val dependencyCount = dependencyCount()
        if (dependencyCount < subDependencyElement.size) {
            return false
        }
        for (i in subDependencyElement.indices) {
            if (getSubDependencyElement(i) != subDependencyElement[i]) {
                return false
            }
        }
        return true
    }
}